meinberlin/apps/kiezradar/serializers.py

```diff
@@ -0,0 +1,104 @@
+from meinberlin.apps.kiezradar.models import ProjectType
+from meinberlin.apps.kiezradar.models import SearchProfile
+from meinberlin.apps.organisations.models import Organisation
+
+
+class SearchProfileSerializer(serializers.ModelSerializer):
+    """Serializer for the SearchProfile model."""
+
+    query = serializers.PrimaryKeyRelatedField(
+        queryset=KiezradarQuery.objects.all(), allow_null=True, required=False
+    )
+    organisations = serializers.PrimaryKeyRelatedField(
+        queryset=Organisation.objects.all(), many=True, allow_null=True, required=False
+    )
+    districts = serializers.PrimaryKeyRelatedField(
+        queryset=AdministrativeDistrict.objects.all(),
+        many=True,
+        allow_null=True,
+        required=False,
+    )
+    project_types = serializers.PrimaryKeyRelatedField(
+        queryset=ProjectType.objects.all(), many=True, allow_null=True, required=False
+    )
+    topics = serializers.PrimaryKeyRelatedField(
+        queryset=Topic.objects.all(), many=True, allow_null=True, required=False
+    )
+    status_display = serializers.CharField(source="get_status_display", read_only=True)
+
+    class Meta:
+        model = SearchProfile
+        fields = [
+            "id",
+            "user",
+            "name",
+            "description",
+            "disabled",
+            "status",
+            "status_display",
+            "query",
+            "organisations",
+            "districts",
+            "project_types",
+            "topics",
+        ]
+
+        read_only_fields = ["user"]
+
+    def create(self, validated_data):
+        # Pop many-to-many fields from validated_data
+        query = validated_data.pop("query", None)
+        organisations_data = validated_data.pop("organisations", None)
+        districts_data = validated_data.pop("districts", None)
+        project_types_data = validated_data.pop("project_types", None)
+        topics_data = validated_data.pop("topics", None)
+
+        search_profile = SearchProfile.objects.create(**validated_data)
+        if query:
+            query, _ = KiezradarQuery.objects.get_or_create(pk=query.id)
+            search_profile.query = query
+            search_profile.save()
+        if organisations_data:
+            search_profile.organisations.set(organisations_data)
+        if districts_data:
+            search_profile.districts.set(districts_data)
+        if project_types_data:
+            for project_type in project_types_data:
+                project_type, _ = ProjectType.objects.get_or_create(pk=project_type.id)
+                search_profile.project_types.add(project_type)
+        if topics_data:
+            search_profile.topics.set(topics_data)
+
+        return search_profile
+
+    def update(self, instance, validated_data):
+        """
+        Custom update to handle many-to-many relationships.
+        """
+        # Pop many-to-many fields from validated_data
```

Comment:
what happens if I change the `query` on an existing search profile? Is this currently missing here? Do we need a `get_or_create` here as well?

Reply:
no, because the query relation to the search profile is not m2m, but one to many. so a query can linked to many profiles. I ditched the m2m here bcz having many queries for a profile would result in too many results, and also would be slow to load. We can change later if there is a strong request/opinion by the PM

Reply:
In the end, I had to add it in the update() as well, in case the user adds a new query text.